### PR TITLE
add cypress-audits to execute lighthouse during end-2-end tests

### LINKED ISSUE #14615: Add Lighthouse to Cypress tests to identify performance regressions
##### **Overview of the feature request**

This looks really cool. I think we should investigate adding it to JHipster.

[Using Cypress and Google Lighthouse for Performance Testing](https://applitools.com/blog/using-cypress-google-lighthouse-performance-testing)

##### **Motivation for or Use Case**

It'd be nice to know if we've done things that cause worse performance, especially for PWAs.

##### **Related issues or PR**

None.

### PR DESCRIPTION
**Pull request by atomfrede**
Right now the cypress audit uses desktop emulation and no cpu slowdown. With 4x cpu slowdown I get very bad results with angular, so I am not sure what happens in our CI with the current thresholds, especially when the frontend is not packaged in prod mode. Maybe we can add an environment variable to skip audits (e.g. when running in dev mode).

Closes #14615 

---

Please make sure the below checklist is followed for Pull Requests.

- [ ] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [ ] Tests are added where necessary
- [ ] [jhipster-online](https://github.com/jhipster/jhipster-online) was updated if necessary
- [ ] Documentation is added/updated where necessary
- [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (bellow reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

**Comment by atomfrede** [comment]
Looks like some configs passes the audit some not (guess somehow depends on the shared runner). I think it would be best to have a dedicated command e.g. npm run audits which just executes the audit spec while the default cypress run does not.

File: generators/client/templates/angular/package.json.ejs [hunk 1]
```diff
@@ -87,6 +87,8 @@
 <%_ } _%>
 <%_ if (cypressTests) { _%>
     "cypress": "<%= dependabotPackageJson.devDependencies['cypress'] %>",
+    "cypress-audit": "<%= dependabotPackageJson.devDependencies['cypress-audit'] %>",
+    "lighthouse": "<%= dependabotPackageJson.devDependencies['lighthouse'] %>",
```
Comment (mshima) [inline generators/client/templates/angular/package.json.ejs:91]: Inject the dependency like we did at cypress-coverage:
https://github.com/jhipster/generator-jhipster/blob/6b532305ee32497e6a0ed30c3f1140b34d520d99/generators/cypress/index.js#L144-L160

No need to change at every client framework package.json, even blueprints.

File: generators/cypress/templates/src/test/javascript/cypress/plugins/index.ts.ejs [hunk 2]
```diff
@@ -38,54 +41,61 @@ module.exports = (on, config) => {
...
   });
-<%_ if (authenticationTypeOauth2) { _%>
+
   on('task', {
+    lighthouse: lighthouse((lighthouseReport) => {
+      !fs.existsSync('<%= BUILD_DIR %>cypress') && fs.mkdirSync('<%= BUILD_DIR %>cypress', { recursive: true }); 
+      fs.writeFileSync('<%= BUILD_DIR %>cypress/lhreport.html', ReportGenerator.generateReport(lighthouseReport.lhr, 'html')); 
+    }),
+    pa11y: pa11y(),
+    <%_ if (authenticationTypeOauth2) { _%>
```
Comment (mshima) [inline generators/cypress/templates/src/test/javascript/cypress/plugins/index.ts.ejs:57]: ejs indentation is cleaner if don't follow the generated source indentation.
```suggestion
<%_ if (authenticationTypeOauth2) { _%>
```

File: generators/cypress/templates/src/test/javascript/cypress/plugins/index.ts.ejs [hunk 3]
```diff
@@ -38,54 +41,61 @@ module.exports = (on, config) => {
...
     login({ baseUrl, username, password }) {
       return (async () => {
         const browser = await puppeteer.launch({
           args: ['--no-sandbox', '--disable-setuid-sandbox'],
           devtools: false,
           ignoreHTTPSErrors: true,
           headless: true });
-
+
```
Comment (mshima) [inline generators/cypress/templates/src/test/javascript/cypress/plugins/index.ts.ejs:65]: Some lines with extra blank space.
```suggestion

```

**Review by mshima — COMMENTED** [review]
The results seem to change a lot each run.
That may be related to virtual environment shared resources.

So IMO it should be optional and disabled by default.
A test enabling it should be discussed separately, probably an application with no entity with focus on pwa.

**Comment by mshima** [comment]
> Looks like some configs passes the audit some not (guess somehow depends on the shared runner). I think it would be best to have a dedicated command e.g. npm run audits which just executes the audit spec while the default cypress run does not.

A custom command would be great too.

**Comment by atomfrede** [comment]
> The results seem to change a lot each run.
> That may be related to virtual environment shared resources.
> 
> So IMO it should be optional and disabled by default.
> A test enabling it should be discussed separately, probably an application with no entity with focus on pwa.

I did not expect such results, but yes I will disable it by default and it must be started e.g. with `npm run e2e:audits` or so. Maybe separate the specs also by name would make it easier.

File: generators/cypress/templates/src/test/javascript/cypress/integration/audits.spec.ts.ejs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+-%>
+describe('Audits', () => {
+    beforeEach(() => { 
+      cy.visit('/');
```
Comment (vishal423) [inline generators/cypress/templates/src/test/javascript/cypress/integration/audits.spec.ts.ejs:21]: It seems that we are trying to calculate the lighthouse score for the application landing page (`/` route). For more comprehensive coverage, we should do that for each route in our application.

Comment (atomfrede) [reply generators/cypress/templates/src/test/javascript/cypress/integration/audits.spec.ts.ejs:21]: Do you expect huge differences between them? But of course that could be extended.

Comment (vishal423) [reply generators/cypress/templates/src/test/javascript/cypress/integration/audits.spec.ts.ejs:21]: TBH, I don't know. I haven't looked into cypress audits and it's capabilities. Do you have assertions to check on best practices? If yes, then, you may want to run the audit for each page.

In the `svelte` blueprint, I used a different approach and the integration is only in the CI (not in generated code). You can of course do that in the generated code, but that was not the concern that I want to handle at moment.

**Comment by atomfrede** [comment]
I have changed it such that 

* audits are in `.audits.ts`
* dedicated cypress config only executing `*.audits.ts` files
* dedicated commands, e.g. `cypress:audits` which opens cypress with the audits configuration
* also updated the readme which did not contain anything about e2e tests when using cypress yet

File: generators/client/templates/angular/package.json.ejs [hunk 5]
```diff
@@ -161,7 +161,7 @@
     <%_ if (cypressTests) { _%>
     "e2e:cypress:headless": "<%= clientPackageManager %> run e2e:cypress -- --headless",
     "e2e:cypress": "cypress run --browser chrome --record ${CYPRESS_ENABLE_RECORD:-false}",
-    "cypress": "cypress open",
+    "cypress": "cypress open"
```
Comment (atomfrede) [inline generators/client/templates/angular/package.json.ejs:164]:
```suggestion
    "cypress": "cypress open",
```

File: generators/cypress/templates/src/test/javascript/cypress/plugins/index.ts.ejs [hunk 6]
```diff
@@ -38,54 +41,60 @@ module.exports = (on, config) => {
...
       return (async () => {
         const browser = await puppeteer.launch({
           args: ['--no-sandbox', '--disable-setuid-sandbox'],
           devtools: false,
           ignoreHTTPSErrors: true,
           headless: true });
-
         const page = await browser.newPage();
-
+
```
Comment (mshima) [inline generators/cypress/templates/src/test/javascript/cypress/plugins/index.ts.ejs:66]:
```suggestion

```

File: generators/cypress/templates/src/test/javascript/cypress/plugins/index.ts.ejs [hunk 7]
```diff
@@ -38,54 +41,60 @@ module.exports = (on, config) => {
...
           headless: true });
-
         const page = await browser.newPage();
-
+  
         await page.goto(`${baseUrl}oauth2/authorization/oidc`, { // The app redirects to the login-page
           waitUntil: 'networkidle2' // Wait until login-page has been reached
         });
-
+
```
Comment (mshima) [inline generators/cypress/templates/src/test/javascript/cypress/plugins/index.ts.ejs:70]:
```suggestion

```

File: generators/cypress/templates/src/test/javascript/cypress/plugins/index.ts.ejs [hunk 8]
```diff
@@ -38,54 +41,60 @@ module.exports = (on, config) => {
...
-
+  
         await page.waitForSelector('[name="username"]');
         await page.waitForSelector('[name="password"]');
         await page.waitForSelector('input[type="submit"]');
         await page.type('[name="username"]', username);
         await page.type('[name="password"]', password);
         await page.click('input[type="submit"]');
-
+
```
Comment (mshima) [inline generators/cypress/templates/src/test/javascript/cypress/plugins/index.ts.ejs:77]:
```suggestion

```

File: generators/cypress/templates/src/test/javascript/cypress/plugins/index.ts.ejs [hunk 9]
```diff
@@ -38,54 +41,60 @@ module.exports = (on, config) => {
...
         await page.waitForSelector('[name="password"]');
         await page.waitForSelector('input[type="submit"]');
         await page.type('[name="username"]', username);
         await page.type('[name="password"]', password);
         await page.click('input[type="submit"]');
-
+  
         await page.waitForNavigation({ waitUntil: 'networkidle2' }); // Wait until redirected back to the app
-
+
```
Comment (mshima) [inline generators/cypress/templates/src/test/javascript/cypress/plugins/index.ts.ejs:79]:
```suggestion

```

File: generators/cypress/templates/src/test/javascript/cypress/plugins/index.ts.ejs [hunk 10]
```diff
@@ -38,54 +41,60 @@ module.exports = (on, config) => {
...
         await page.type('[name="password"]', password);
         await page.click('input[type="submit"]');
-
+  
         await page.waitForNavigation({ waitUntil: 'networkidle2' }); // Wait until redirected back to the app
-
+  
         const authCookies = await page.cookies();
-
+
```
Comment (mshima) [inline generators/cypress/templates/src/test/javascript/cypress/plugins/index.ts.ejs:81]:
```suggestion

```

File: generators/cypress/templates/src/test/javascript/cypress/plugins/index.ts.ejs [hunk 11]
```diff
@@ -38,54 +41,60 @@ module.exports = (on, config) => {
...
-    }
+        const browser = await puppeteer.launch({ ignoreHTTPSErrors: true });
+        const page = await browser.newPage();
+        await page.goto(`${baseUrl}api/logout`, {
+          waitUntil: 'networkidle2'
+        });
+        return true;
+      })
+    } 
+    <%_ } _%>
```
Comment (mshima) [inline generators/cypress/templates/src/test/javascript/cypress/plugins/index.ts.ejs:96]:
```suggestion
<%_ } _%>
```

File: generators/cypress/templates/cypress-audits.json.ejs [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+    "integrationFolder": "<%= CLIENT_TEST_SRC_DIR %>cypress/integration",
+    "fixturesFolder": "<%= CLIENT_TEST_SRC_DIR %>cypress/fixtures",
+    "pluginsFile": "<%= CLIENT_TEST_SRC_DIR %>cypress/plugins/index.ts",
+    "screenshotsFolder": "<%= BUILD_DIR %>cypress/screenshots",
+    "chromeWebSecurity": <% if(authenticationTypeOauth2) { %>false<% } else { %>true<% } %>,
+    "viewportWidth": 1200,
+    "viewportHeight": 720,
+    "retries": 2
+  }
+
```
Comment (mshima) [inline generators/cypress/templates/cypress-audits.json.ejs:33]:
```suggestion

```

File: generators/cypress/templates/src/test/javascript/cypress/integration/lighthouse.audits.ts.ejs [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+        // Change the CPU slowdown multiplier to emulate different kind of devices
+        // See https://github.com/GoogleChrome/lighthouse/blob/master/docs/throttling.md#cpu-throttling for details
+        throttling: {
+          cpuSlowdownMultiplier: 1
+        },
+        screenEmulation: {disabled: true}
+      };
+      cy.lighthouse(customThresholds, desktopConfig);
+    });
+});
```
Comment (mshima) [inline generators/cypress/templates/src/test/javascript/cypress/integration/lighthouse.audits.ts.ejs:46]:
```suggestion
});

```

**Review by mshima — APPROVED** [review]
I tried to remove whitespaces, but removed lines.

LGTM.

**Comment by atomfrede** [comment]
Thanks @mshima for your review. you are awesome!

**Comment by atomfrede** [comment]
Bounty claimed https://opencollective.com/generator-jhipster/expenses/40724

**Comment by pascalgrimaud** [comment]
@atomfrede : approved

**Pull request merged**